Log Management

aluzzardi opened the pull request:
This change adds support for log management in the manager, agent and
CLI.

The log broker is currently naive and broadcasts subscriptions to all
agents, which in turn need to perform filtering to figure out if the
subscription is relevant to them.

In the future, the broker should be smarter and dispatch subscriptions
only to concerned agents.

The basic logging functionality works.

Fixes #1332

/cc @stevvooe @aaronlehmann @diogomonica

aaronlehmann commented on agent/worker.go:
```diff
@@ -39,24 +42,26 @@ type statusReporterKey struct {
-	listeners map[*statusReporterKey]struct{}
-	secrets   *secrets
+	db                *bolt.DB
+	taskevents        *events.Broadcaster
```
This needs to be closed once it is no longer being used, otherwise it will leak a goroutine.

aluzzardi replied:
I've updated the LogBroker - however, there's no way to terminate a Worker so it should never leak

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -15,58 +19,152 @@ import (
 	return &LogBroker{
-		broadcaster: events.NewBroadcaster(),
+		broadcaster:   events.NewBroadcaster(),
+		subscriptions: events.NewBroadcaster(),
```
Same with these - they need to be closed when the log broker stops.

Take care when closing them that there are no sinks still attached to the broadcasters; otherwise it can deadlock. We have a wrapper in `manager/state/watch` that we use to protect against this, which you may find useful here.

aaronlehmann commented on agent/worker.go:
```diff
@@ -376,3 +382,69 @@ func (w *worker) updateTaskStatus(ctx context.Context, tx *bolt.Tx, taskID strin
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	log.G(ctx).Infof("Received subscription %s", subscription.ID)
```
It's probably a good idea to log the selectors as well.

aaronlehmann commented on agent/exec/container/controller.go:
```diff
@@ -395,7 +395,65 @@ func (r *controller) Remove(ctx context.Context) error {
+	}
+
+	ctx, cancel := context.WithCancel(pctx)
+	defer cancel()
```
What's the cancel useful for? Do any of the function calls below spawn goroutines that outlive this function?

aluzzardi replied:
Removed

aaronlehmann replied:
> Removed

The `WithCancel` still seems to be in the code

aaronlehmann commented on agent/worker.go:
```diff
@@ -376,3 +382,69 @@ func (w *worker) updateTaskStatus(ctx context.Context, tx *bolt.Tx, taskID strin
+
+func (w *worker) Subscribe(ctx context.Context, subscription *api.SubscriptionMessage) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
What's the cancel useful for? Do any of the function calls below spawn goroutines that outlive this function?

aluzzardi replied:
Right - removed the sub-context

aaronlehmann commented on agent/exec/container/controller.go:
```diff
@@ -388,6 +395,173 @@ func (r *controller) Remove(ctx context.Context) error {
+
+			Data: parts[1],
+		}, false); err != nil {
+			return errors.Wrap(err, "failed publisher log message")
```
"failed to publish log message"?

stevvooe commented on agent/exec/controller.go:
```diff
@@ -45,6 +45,33 @@ type Controller interface {
+type LogPublisherFunc func(ctx context.Context, message api.LogMessage, close bool) error
+
+// Publish calls the wrapped function.
+func (fn LogPublisherFunc) Publish(ctx context.Context, message api.LogMessage, close bool) error {
```
No need to propagate `close` here. Just cancel the context.

stevvooe commented on agent/exec/container/controller.go:
```diff
@@ -388,6 +395,173 @@ func (r *controller) Remove(ctx context.Context) error {
+	for {
+		// so, message header is 8 bytes, treat as uint64, pull stream off MSB
+		var header uint64
+		if err := binary.Read(brd, binary.BigEndian, &header); err != nil {
```
Most of this should be moved into the adapter.

aluzzardi commented on agent/exec/container/controller_integration_test.go:
```diff
@@ -69,3 +91,9 @@ func TestControllerFlowIntegration(t *testing.T) {
 	}
 }
+
+type logPublisherFn func(ctx context.Context, message api.LogMessage, close bool) error
```
Exec defines a log publish, not log publish**er**

aluzzardi replied:
Never mind, this is old code

aluzzardi commented:
@aaronlehmann @stevvooe PTAL

aaronlehmann commented:
> I've updated the LogBroker - however, there's no way to terminate a Worker so it should never leak

I don't understand. Doesn't `docker swarm leave` terminate the worker?

aaronlehmann commented:
@aluzzardi: Unit tests seemed to hang in CI.

aluzzardi commented:
> I don't understand. Doesn't docker swarm leave terminate the worker?

It just stops the agent

aluzzardi commented:
> @aluzzardi: Unit tests seemed to hang in CI.

`make test` works locally :/ I'll try `make ci`

aluzzardi commented:
Rebased.

PTAL

stevvooe commented on agent/agent.go:
```diff
@@ -186,6 +186,22 @@ func (a *Agent) run(ctx context.Context) {
 			}
+		case sub := <-session.subscriptions:
+			if sub.Close {
+				if cancel, ok := subscriptions[sub.ID]; ok {
```
Does this ever get removed from the map?

aaronlehmann commented:
Needs a rebase :)

aaronlehmann commented on agent/exec/container/controller.go:
```diff
@@ -388,6 +392,132 @@ func (r *controller) Remove(ctx context.Context) error {
+	var (
+		// use a rate limiter to keep things under control but also provides some
+		// ability coalesce messages.
+		limiter = rate.NewLimiter(rate.Every(time.Second), 10<<20) // 10 MB/s
```
Probably something for the future, but this rate limit seems like something we should make configurable.

aaronlehmann commented on agent/worker.go:
```diff
@@ -375,3 +383,64 @@ func (w *worker) updateTaskStatus(ctx context.Context, tx *bolt.Tx, taskID strin
+		}
+
+		return false
+	}
```
Maybe a future optimization: we could build maps to avoid all this iteration.

aaronlehmann commented on manager/manager.go:
```diff
@@ -324,6 +330,8 @@ func (m *Manager) Run(parent context.Context) error {
 	api.RegisterHealthServer(m.server, authenticatedHealthAPI)
 	api.RegisterRaftMembershipServer(m.server, proxyRaftMembershipAPI)
 	api.RegisterControlServer(m.server, authenticatedControlAPI)
+	api.RegisterLogsServer(m.localserver, localProxyLogsAPI)
```
Can you please move this line below to group it with the other calls that register services on `m.localserver`?

aaronlehmann commented on manager/manager.go:
```diff
@@ -225,6 +227,7 @@ func New(config *Config) (*Manager, error) {
 		listeners:   listeners,
 		caserver:    ca.NewServer(raftNode.MemoryStore(), config.SecurityConfig),
 		dispatcher:  dispatcher.New(raftNode, dispatcherConfig),
+		logbroker:   logbroker.New(),
```
Should `logbroker` be started when we become the leader, and stop when we lose leadership? I'd think we want to close log streams when we lose leadership, and refuse to open new ones.

What happens right now when there's a leader failover? Do the agents keep publishing logs to the old leader as long as existing streams are open, or does the raft proxy direct these logs to the new leader instead? Will the client be able to automatically reestablish a stream on leader failover?

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// PublishLogs publishes log messages for a given subscription
+func (lb *LogBroker) PublishLogs(ctx context.Context, request *api.PublishLogsRequest) (*api.PublishLogsResponse, error) {
```
What's the authorization model for this RPC to prevent malicious workers from injecting logs into another node's stream? Is the idea that `SubscriptionID` is like a secret token?

It probably wouldn't hurt to get the node ID from `ca.RemoteNode(ctx)` and verify that this particular node was either selected directly in the `LogSelector` or has a task or service that was selected. We should also verify that the node ID given in the `LogContext` inside `request` matches the sender's node ID, that the task ID in `LogContext` is assigned to this node, and that the service ID in `LogContext` matches the task's service.

cc @diogomonica

aluzzardi replied:
I wanted to avoid using the store at this point.

In a next iteration, there's going to be a subscription watch model where we expand services and lookup tasks to figure out on which nodes the subscription should land.

At that point, it would be a trivial map lookup to see if a given subscription is active on a given node.

aluzzardi replied:
What I'm going to do in the meantime is this:

Every log message contains a `Context`, which contains the `NodeID`. I'll enforce at the publish level that the NodeID in the context matches the remote NodeID.

This means that even if a node manages to publish logs for another subscription, it **can't** impersonate the other node. Those logs will be clearly marked with the source.

aaronlehmann commented on api/logbroker.proto:
```diff
@@ -0,0 +1,174 @@
+
+	// PublishLogs receives sets of log messages destined for a single
+	// subscription identifier.
+	rpc PublishLogs(PublishLogsRequest) returns (PublishLogsResponse) {
```
Any thoughts about making this a client streaming RPC instead of a one-off RPC? I don't have a good sense of the overhead that each kind of RPC has, but I'd imagine a node will typically publish many messages as part of a log subscription, and by using a stream instead of many RPC invocations, we'd lower the cost of the things like the TLS authorization checks that have to run every time the RPC is invoked.

aluzzardi replied:
Makes sense. @stevvooe ?

aaronlehmann commented:
Unit tests are getting stuck again.

aluzzardi commented:
@aaronlehmann Yeah. It should be fixed now. I did a little bit of what I'll improve later: now the broker keeps track of all active subscriptions and when a new agent connects, we send down the active list before the new ones in a non-racy way.

Removed the Sleep from the tests and added an additional test

PTAL

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -54,6 +60,38 @@ func validateSelector(selector *api.LogSelector) error {
 
+func (lb *LogBroker) registerSubscription(subscription *api.SubscriptionMessage) {
+	lb.mu.Lock()
+	defer lb.mu.Unlock()
```
Probably better to release the lock before calling `Publish`.

I might change the name of the mutex to make it clear it protects `registeredSubscriptions` specifically.

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -54,6 +60,38 @@ func validateSelector(selector *api.LogSelector) error {
+	subscription.Close = true
+
+	lb.mu.Lock()
+	defer lb.mu.Unlock()
```
Similar to above, can release the lock before calling `Publish`.

aaronlehmann replied:
Seems fine. Clearly I misunderstood the purpose of this lock.

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -54,6 +60,38 @@ func validateSelector(selector *api.LogSelector) error {
+
+func (lb *LogBroker) subscriptions() ([]*api.SubscriptionMessage, chan events.Event, func()) {
+	lb.mu.RLock()
+	defer lb.mu.RUnlock()
```
and before calling `Watch`.

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -124,10 +159,28 @@ func (lb *LogBroker) ListenSubscriptions(request *api.ListenSubscriptionsRequest
 
 	log.Debug("node registered")
+
+	// Start by sending down all active subscriptions.
```
Can you add a `TODO` about filtering the subscriptions relevant to each node?

aluzzardi commented:
Addressed most comments except start/stop the broker on leadership

aaronlehmann commented:
@aluzzardi: The hanging test problem seems to be back.

aluzzardi commented:
@aaronlehmann Done - I've fixed the hanging test (test's fault - it was not setting the correct NodeID and got rejected), added a Start/Stop on leadership changes and also fixed the Raft proxying part.

For the latter part, would you mind double checking what I'm doing? Specifically on the security side (I'm mounting the log server on TCP so it can be proxied by non-leaders):
https://github.com/docker/swarmkit/pull/1679/commits/965d2f0e6414b12ff2732bf86f3f8b31ed8c9e40

aaronlehmann commented:
lint error

aaronlehmann commented:
The change related to raft proxying looks right. Before, `LogsServer` was set up to proxy incoming requests from the control socket to the leader, but there was no `LogsServer` bound to the TCP server, so it wouldn't have been possible for the leader to actually receive these requests over the network. I assume all `LogsServer` and `LogBrokerServer` requests are supposed to be handled by the leader?

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -41,11 +41,32 @@ func New() *LogBroker {
+	lb.mu.Lock()
+	lb.stopped = make(chan struct{})
+	lb.logQueue = watch.NewQueue()
+	lb.subscriptionQueue = watch.NewQueue()
```
If you create the queues here, you shouldn't create them in `New`. Otherwise there's a leak.

aluzzardi replied:
Thanks, I missed this!

Fixed

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -41,11 +41,32 @@ func New() *LogBroker {
+
 // Stop stops the log broker
-func (lb *LogBroker) Stop() {
+func (lb *LogBroker) Stop() error {
```
`Stop` should wait for all running RPCs to stop (probably using a waitgroup).

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -41,11 +41,32 @@ func New() *LogBroker {
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
```
If `LogBroker` doesn't have a main loop, maybe it doesn't make sense to have a blocking `Run` function that only waits for `Stop` to be called. The manager could just call `New` whenever it becomes the leader. But I can see the argument for consistency. On the other hand, I dislike goroutines that do nothing.

aaronlehmann commented on manager/manager.go:
```diff
@@ -712,6 +729,7 @@ func (m *Manager) becomeLeader(ctx context.Context) {
 // becomeFollower shuts down the subsystems that are only run by the leader.
 func (m *Manager) becomeFollower() {
 	m.dispatcher.Stop()
+	m.logbroker.Stop()
```
It looks like calling `m.logbroker.Stop` multiple times will panic. Since both `becomeFollower` and `(*Manager).Stop` can call `m.logbroker.Stop`, either logbroker's `Stop` function should protect against this (as I believe most `Stop` functions do), or the manager should set `m.logbroker` to `nil` after calling `Stop` and check before each use.

aluzzardi commented:
@aaronlehmann Thanks! Agreed on all counts, will make the changes

aaronlehmann commented:
Oh, interesting, CI found a race. `SubscribeLogs` uses the queue created in `Run` without locking. `Stop` should probably wait for all outstanding RPCs to terminate before returning, and RPCs should error out if the log broker is not currently running. I think I mentioned this before but maybe the comment got lost.

aluzzardi commented:
@aaronlehmann Yeah, right, I was looking into that now.

I'm seeing a similar pattern in dispatcher as well (check if running on every RPC).

I'll adopt the same pattern but I'm wondering if that's repeated work. Shouldn't the manager basically take care of 1) stop routing traffic to the service 2) call stop and do the opposite on promotion? That way our services could be "dumber" (in terms of init/deinit logic)

Building on that, we could actually have a `Servicer` (bad name, I know) interface with `Run` & `Stop` and the manager could simply keep a list of services and perform the Run + Mount on startup/promotion and Umount + Stop on demotion. Then adding a new service such as the broker would just be a matter of appending the object to the list.

aluzzardi commented:
Regarding the above - I'm also wondering if the broker really needs a `Stop` function. Wouldn't it be simpler if the demotion code just stopped the actual server altogether, cancelling all active connections?

aaronlehmann commented:
How do you stop the server without killing the whole GRPC listener?

aaronlehmann commented:
> I'll adopt the same pattern but I'm wondering if that's repeated work. Shouldn't the manager basically take care of 1) stop routing traffic to the service 2) call stop and do the opposite on promotion? That way our services could be "dumber" (in terms of init/deinit logic)

Maybe, but it's not easy. (1) probably involves some codegenned wrappers, unless there's some way I don't know of to "stop traffic" for a service (there is no `Unregister` function).

aluzzardi commented:
Just pushed an update addressing the remaining comments (I only ran `make test` so other things may fail, I'll fix that tomorrow).

There's a bunch of different changes but the most relevant are:

> If worker creates a broadcaster, it needs to have a well-defined lifecycle so the broadcaster's resources can be freed. There are some real-world cases in Docker where multiple workers would be created without a restart in between; for example after encryption-at-rest lands each attempt to unlock the manager will create a worker.

@aaronlehmann Added stop safeguards in https://github.com/docker/swarmkit/pull/1679/commits/9a03850b1a7bfc02a9501b454f2f43bc48678e79

> https://github.com/docker/swarmkit/pull/1679/files#r85861157 needs to be addressed

This is about the agent leaking resources (worker event.Broadcaster) because we don't cleanly shut down the worker.

I've changed the worker interface to add a `Close` function (https://github.com/docker/swarmkit/pull/1679/commits/75a4aa5ce9bfc43d5caaf910c6b2d7532298249b) and also moved the manager/state/watch into its own package (https://github.com/docker/swarmkit/pull/1679/commits/ab301dfb43c109dabf998774a06e44608a5a2713).

@stevvooe Can you verify if the worker interface and agent integration look alright?
@aaronlehmann Could you double check if the leaks are gone?

> I'd like https://github.com/docker/swarmkit/pull/1679/files#r85228089 to be considered before merging.

That's about logs being unary RPCs rather than streams.

I've implemented that in https://github.com/docker/swarmkit/pull/1679/commits/49e48bec3d82ced146c1cef5347a935f5cdb2e40

@stevvooe @aaronlehmann Are you happy with how it looks?
@LK4D4 I had to change the codegen wrapper to support client streaming. Turns out, the interface for both Client and Client&Server is the same. Could you check if it looks alright?

aaronlehmann commented on manager/logbroker/broker_test.go:
```diff
@@ -228,7 +235,13 @@ func testLogBrokerEnv(t *testing.T) (context.Context, *LogBroker, *ca.SecurityCo
 
+	go func() {
+		broker.Run(ctx)
+	}()
```
nit: Don't need a closure.

aaronlehmann commented on manager/logbroker/broker.go:
```diff
@@ -28,24 +34,51 @@ type LogBroker struct {
 	lb.logQueue.Close()
 	lb.subscriptionQueue.Close()
+
+	return nil
```
I think it's good practice for `Stop` to block until all current RPCs are finished. I don't see anything especially dangerous in the RPCs like interactions with Raft, but we may add something in the future. Also, right now it's theoretically possible for an RPC started before a leader reelection to interact with the new queue created by `Run`, which feels wrong.

aluzzardi replied:
I followed the same pattern as the dispatcher and keymanager - neither of them has a blocking `Stop` (and they do interact with Raft).

Do you want me to make stop blocking or keeping it consistent with the rest?

aaronlehmann commented on api/logbroker.proto:
```diff
@@ -94,14 +94,14 @@ service Logs {
 
 message SubscribeLogsMessage {
-	repeated LogMessage messages = 1 [(gogoproto.nullable) = false];
+	LogMessage message = 1 [(gogoproto.nullable) = false];
```
It still seems useful for this to be a repeated field so we can send down multiple messages at a time if we later decide to. That will probably cut down some framing overhead.

aluzzardi commented:
@aaronlehmann Addressed your comments except for blocking Stop (waiting your response on whether we should do it or not)

aaronlehmann commented:
Sorry for the misinformation about blocking `Stop`. I was sure that dispatcher did that way. Turns out it was changed here: 6ec2d00201a1ca50cafb6d19823f7863e4dc81f5

Let's keep `Stop` non-blocking for now, to be consistent with dispatcher and ca.

stevvooe commented on agent/exec/container/adapter.go:
```diff
@@ -258,6 +258,49 @@ func (c *containerAdapter) createVolumes(ctx context.Context) error {
+	apiOptions := types.ContainerLogsOptions{
+		Follow: options.Follow,
+
+		// TODO(stevvooe): Parse timestamp out of message. This
```
This comment is no longer valid. We parse below. :)

Sorry for not removing.
